n2v/methods/oucarter.py

```diff
@@ -0,0 +1,351 @@
+"""
+oucarter.py
+
+Functions associated with Ou-Carter inversion
+"""
+
+import numpy as np
+from opt_einsum import contract
+import psi4
+import time
+
+class OC():
+    """
+    Ou-Carter density to potential inversion [1].
+    [1] [J. Chem. Theory Comput. 2018, 14, 5680−5689]
+    """
+
+    def _get_l_kinetic_energy_density_directly(self, D, C, grid_info=None):
+        """
+        Calculate $\frac{\tau_L^{KS}}{\rho^{KS}}-\frac{\tau_P^{KS}}{\rho^{KS}}$,
+        (i.e. the 2dn and 3rd term in eqn. (17) in [1]):
+        """
+
+        if grid_info is None:
+            tauLmP_rho = np.zeros(self.Vpot.grid().npoints())
+            nblocks = self.Vpot.nblocks()
+
+            points_func = self.Vpot.properties()[0]
+            blocks = None
+
+        else:
+            blocks, npoints, points_func = grid_info
+            tauLmP_rho = np.zeros(npoints)
+            nblocks = len(blocks)
+
+        points_func.set_deriv(2)
+
+        iw = 0
+        for l_block in range(nblocks):
+            # Obtain general grid information
+            if blocks is None:
+                l_grid = self.Vpot.get_block(l_block)
+            else:
+                l_grid = blocks[l_block]
+            l_npoints = l_grid.npoints()
+
+            points_func.compute_points(l_grid)
+            l_lpos = np.array(l_grid.functions_local_to_global())
+            if len(l_lpos) == 0:
+                iw += l_npoints
+                continue
+            l_phi = np.array(points_func.basis_values()["PHI"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_x = np.array(points_func.basis_values()["PHI_X"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_y = np.array(points_func.basis_values()["PHI_Y"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_z = np.array(points_func.basis_values()["PHI_Z"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_xx = np.array(points_func.basis_values()["PHI_XX"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_yy = np.array(points_func.basis_values()["PHI_YY"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_zz = np.array(points_func.basis_values()["PHI_ZZ"])[:l_npoints, :l_lpos.shape[0]]
+
+            lD = D[(l_lpos[:, None], l_lpos)]
+            # lC = C[l_lpos, :]
+
+            rho = contract('pm,mn,pn->p', l_phi, lD, l_phi)
+            rho_inv = 1/rho
+
+            # Calculate the second term 0.25*\nabla^2\rho
+            laplace_rho_temp = contract('ab,pa,pb->p', lD, l_phi, l_phi_xx + l_phi_yy + l_phi_zz)
```

Comment:
Can we have a latex comment for this It would help a lot to read through the code. I understand that the gradient is expressed as:

$ \nabla**2 n(r) = \sum_{\mu \nu} P_{\mu \nu} \Phi_{mu} \nabla**2 \Phi_{\nu} + 2 * \nabla \Phi_{\mu} \nabla \Phi_{\nu} + \Phi_{\nu} \nabla**2 \Phi_{\mu} $

Is this an analogous expression?

Reply:
Will do.